### ISSUE TITLE
Template for default.nix

### ISSUE THREAD

**Issue by anton-trunov** [coq-community/templates#9]
To fix build on Coq 8.9 and Coq 8.10 for https://github.com/coq-community/coq-bits, this needed to be added to `default.nix`:

```
installPhase = ''
    make -f CoqMakefile COQLIB=$out/lib/coq/${coq.coq-version}/ install
  '';
```

See https://github.com/coq-community/coq-bits/pull/3

**Comment by palmskog** [comment]
I have no idea if this is the right way to do it, I think we need someone with Nix expertise to comment. I thought it always sufficed to have `make install` to support Nix fully?

cc: @Zimmi48 @vbgl

**Comment by vbgl** [comment]
The usual coq-makefile wants to install third-party libraries next to Coq standard library. This is incompatible with the immutability of the nix store. So indeed, it is common practice to explicitly tell coq-makefile where to install *this* library.

It is also possible to only set `installFlags` to avoid repeating `make` and `install` and co.

**Comment by palmskog** [comment]
@vbgl but I thought the location problem was already handled by the following line in @anton-trunov's `default.nix`:
```
installFlags = "COQLIB=$(out)/lib/coq/${coq.coq-version}/";
```
See https://github.com/coq-community/bits/blob/master/default.nix#L29

I was at least under the impression that `coq_makefile` respected the `COQLIB` variable and that saying explicitly `installPhase = ...` would therefore be redundant.

**Comment by vbgl** [comment]
Thanks for clarifying. Indeed, setting the `COQLIB` variable in the `installFlags` should be enough. Is there a failing example to look at?

**Comment by anton-trunov** [comment]
@vbgl Here you go: https://github.com/coq-community/bits/pull/8/checks?check_run_id=424943595. I removed that snippet and now it builds on Coq 8.7 and 8.8, but fails on 8.9 and 8.10.

**Comment by palmskog** [comment]
So apparently only builds with Coq 8.9 and later are failing, e.g., https://travis-ci.com/coq-community/bits/jobs/283318683

Example:
```
install flags: SHELL=/nix/store/715fljdrspphbdk255r729zxp4w7lcgk-bash-4.4-p23/bin/bash COQLIB=\$\(out\)/lib/coq/8.9/ install
coq_makefile -f _CoqProject -o CoqMakefile
Fatal error: exception Sys_error("/nix/store/gj1li5l0r5c56796f6m3ld72dham244r-bits/lib/coq/8.9//tools/CoqMakefile.in: No such file or directory")
```

I'm thinking this *could* be some weird Makefile-Nix interaction issue. @anton-trunov can you reproduce this error when using another delegating Makefile like this one? https://github.com/coq-community/huffman/blob/master/Makefile

**Comment by anton-trunov** [comment]
@palmskog Indeed you are right about the Chapar Makefile: https://github.com/coq-community/bits/runs/424973800

**Comment by anton-trunov** [comment]
Looks like this should be reported as a Coq issue. Because the Coq reference manual suggests using the original template: https://coq.github.io/doc/v8.11/refman/practical-tools/utilities.html#reusing-extending-the-generated-makefile

**Comment by anton-trunov** [comment]
As pointed out by @vbgl, switching to
```
installFlags = "COQMF_COQLIB=$(out)/lib/coq/${coq.coq-version}/";
```
also solves the issue (this does not require changing `Makefile`)

**Comment by palmskog** [comment]
OK, this is still a very strange error to me, but I guess we can change the `default.nix.mustache` template to replace `COQLIB` with `COQMF_COQLIB`.

**Comment by Zimmi48** [comment]
I haven't followed this issue closely. What does `COQMF_COQLIB` means?

**Comment by vbgl** [comment]
Maybe Enrico can shed some light on this issue.

Ping @gares.

**Comment by gares** [comment]
I don't have time to read all linked material. All the variables part of `Makefile.coq.conf` are prefixed by `COQMF_` so that you can include that makefile (containing only variables coming from coqtop -config roughly) in other makefiles. Internally `Makefile.coq` uses shorter names, eg COQLIB that take as default value what is in `Makefile.coq.conf`.

`Makefile.coq` already provides a variable to "redirect" the installation *root*, and is probably called DESTDIR or PREFIX I don't recall. If you need to override the name `coq` in that path, I suggest you introduce a new variable (and temporarily hijack COQMF_COQLIB, if that works).

But I don't get one thing. If you install stuff in non-standard places, Coq has still to find them.
Are you saying you did not change its paths via ./configure but you are exporting COQLIB in the user shell? In case, please don't do it, since coq users are used to that variable and may tamper with it (even under nix). If you really want to do it via the env add a specific, not for the user, variable.

**Comment by vbgl** [comment]
Thanks for clarifying the difference between `COQLIB` and `COQMF_COQLIB`.

If the purpose `DESTDIR` is what you describe, then it’s probably broken:
~~~
COQLIBINSTALL = $(call concat_path,$(DESTDIR),$(COQLIB)user-contrib)
~~~

Suppose I use a local Coq installation (in `$HOME/local/lib/coq/`) to build a library that I want to install into `/path/usr/lib/coq/user-contrib`; how should I set `DESTDIR`? Something like `/path/usr/` seems reasonable but won’t work…

So, coq-makefile always want to install into `$DESTDIR/$COQLIB`, so overriding `$COQLIB` is mandatory (by design of coq-makefile) if we want to choose the installation directory.

But rest assured that this overriding is local to the call to `make install`.

**Comment by Zimmi48** [comment]
The way Coq finds these libraries is via $COQPATH, which is set when entering a `nix-shell`. This is a pretty common practice across a large variety of languages.

**Comment by gares** [comment]
Yes, I know, *but* while, say, JAVAHOME existed since day one with that meaning, COQLIB has a different meaning. To my understanding was supposed to be used by the user to override the coq default. In some sense it is legitimate to write "whatever COQLIB=this" while in the nix world only "whatever COQLIB+=extra" really makes sense.

If you want to take over COQLIB, I'm fine with it, but you (nix guys) should tell coq users and document it. I think I've had the same complaint on OCAMLPATH with @vgbl (because setting it is supposed to be legit, and in nix is is not, you can only append).

**Comment by vbgl** [comment]
Are you mixing `COQLIB` and `COQPATH`?

The `COQPATH` variable is used as per its [documentation](https://coq.github.io/doc/v8.11/refman/language/gallina-extensions.html?highlight=coqpath#libraries-and-filesystem) in nix packages.

The `COQLIB` variable is some implementation detail of coq-makefile; overriding it while calling `make install` is the only way I know to chose the target directory when installing a library that uses coq-makefile. If there is a better way, please let me know.

**Comment by gares** [comment]
Maybe it is not documented, but Coq uses `COQLIB`
https://github.com/coq/coq/blob/c6e89e2b13d7fa99aaaaf02fc3430ae6c5f00beb/lib/envars.ml#L97
for example we use it in the test suite, sphinx and CI to override where Coq looks for the stdlib, that depends on the build system for example. But yes, in my nix-rant I did confuse it with COQPATH that seems to be legit to use.

I looked at CoqMakefile.in and I believe it is safe if you call "make install" alone, but the variable is also used elsewhere, so if one calls "make all install" (which is wrong in many ways, but how many Coq users know it?) overriding COQLIB make things will go wrong.

So, as I suggested earlier, I prefer you override COQLIB only as a temporary solution and add a new variable, say COQLIB_INSTALL, so that overriding it has only the expected effect (and document it in the refman, since version 2 of coq_makefile I tried to document  the variables that are supposed to be overridden by the user, which is not the totality of variables). Would that work for you?
